postgres/client.js

```diff
@@ -1,39 +1,60 @@
 'use strict';
 
-const bluebird = require('bluebird'),
-  { Client } = require('pg'),
-  QueryStream = require('pg-query-stream'),
-  { POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_HOST, POSTGRES_PORT, POSTGRES_DB } = require('../services/constants'),
-  { PUT, GET, DEL, BATCH, READ_STREAM } = require('./queries'),
+const { POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_HOST, POSTGRES_PORT, POSTGRES_DB } = require('../services/constants'),
   { notFoundError } = require('../services/errors'),
-  { parseOrNot, wrapInObject } = require('../services/utils'),
+  { parseOrNot } = require('../services/utils'),
+  { findSchemaAndTable, wrapJSONStringInObject } = require('../services/utils'),
+  knexLib = require('knex'),
   TransformStream = require('../services/list-transform-stream');
-var client = new Client({
-  user: POSTGRES_USER,
-  host: POSTGRES_HOST,
-  database: POSTGRES_DB,
-  password: POSTGRES_PASSWORD,
-  port: POSTGRES_PORT,
-});
+var knex;
 
 /**
- * Connect to the DB
+ * Connect to the default DB and create the Clay
+ * DB. Once the DB has been made the connection
+ * can be killed and we can try to re-connect
+ * to the Clay DB
  *
- * @return {Promise}
+ * @returns {Promise}
  */
-function connect() {
-  return client.connect();
+function createDBIfNotExists() {
+  const tmpClient = knexLib({
+    client: 'pg',
+    connection: {
+      host: POSTGRES_HOST,
+      user: POSTGRES_USER,
+      password: POSTGRES_PASSWORD,
+      database: 'postgres',
+      port: POSTGRES_PORT
+    }
+  });
+
+  return tmpClient.raw(`CREATE DATABASE ${POSTGRES_DB}`)
+    .then(() => tmpClient.destroy())
+    .then(connect);
 }
 
 /**
- * Execute a query using the client
- * scoped to the module
+ * Connect to the DB. We need to do a quick check
+ * to see if we're actually connected to the Clay
+ * DB. If we aren't, connect to default and then
+ * use that connection to create the Clay DB.
  *
- * @param  {String} query
- * @return {Promise}
+ * @returns {Promise}
  */
-function query(query) {
-  return client.query(query);
+function connect() {
+  knex = knexLib({
+    client: 'pg',
+    connection: {
+      host: POSTGRES_HOST,
+      user: POSTGRES_USER,
+      password: POSTGRES_PASSWORD,
+      database: POSTGRES_DB,
+      port: POSTGRES_PORT
+    }
+  });
+
+  return knex.table('information_schema.tables').first()
+    .catch(createDBIfNotExists);
```

Comment:
I would leave a comment here about improving the catch to make sure that the error thrown is an actual DB not found error - otherwise I think you might end up with an uncaught promise rejection when it tries to create the db.  Here's a sample of an error thrown which is a db not found:

```
{ error: database "none" does not exist
    at Connection.parseE (/redact/node_modules/pg/lib/connection.js:553:11)
    at Connection.parseMessage (/redact/node_modules/pg/lib/connection.js:378:19)
    at Socket.<anonymous> (/redact/node_modules/pg/lib/connection.js:119:22)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:594:20)
  name: 'error',
  length: 89,
  severity: 'FATAL',
  code: '3D000',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'postinit.c',
  line: '846',
  routine: 'InitPostgres' }
```

Reply:
👍